core/transactions/staged_mutation.cxx

```diff
@@ -150,12 +167,14 @@ staged_mutation_queue::iterate(std::function<void(staged_mutation&)> op)
     }
 }
 
-void
-staged_mutation_queue::commit(attempt_context_impl* ctx)
+std::future<void>
+staged_mutation_queue::commit_mutation(attempt_context_impl* ctx, staged_mutation& item, unstaging_state& state)
 {
-    CB_ATTEMPT_CTX_LOG_TRACE(ctx, "staged mutations committing...");
-    std::lock_guard<std::mutex> lock(mutex_);
-    for (auto& item : queue_) {
+    auto barrier = std::make_shared<std::promise<void>>();
+    auto future = barrier->get_future();
+
+    std::async(std::launch::async, [&]{
```

Comment:
Could you give me a brief description of how this (_gestures vaguely at std::async_) works in terms of threads, pools, etc.

Reply:
According to the [docs](https://en.cppreference.com/w/cpp/thread/async) if you specify the `std::launch::async` launch policy to `std::async`, it executes the function in a new thread equivalent to using `std::thread`. Then `std::async` returns a future (so I shouldn't have used the `barrier` promise there and should've actually stored the future that it returns - the updated code I'll push shortly should address this).

Another issue with this piece of code is that if I have N mutations I'm creating N threads immediately and a lot of them will be blocked until they can do the commit/rollback if N > 1000. I will update it to have the main thread (or a thread responsible for starting the threads for each mutation) block until the in flight commits/rollbacks are less than 1000, so only that one thread will be blocked

Reply:
Now using the cluster's IO context instead